README.md
# OS161
Custom built Operating System


CHANGES
OS/161 was written by David A. Holland, with contributions from
	Amos Blackman
	Alexandra Fedorova
	Ada T. Lim
	Georgi Matev
	Jay Moorthi
	Geoffrey Werner-Allen

Additional small patches and bug reports have been contributed by
various other people, all of whom are (hopefully) listed below.

------------------------------------------------------------

20140924 dholland	OS/161 1.99.08 released. Consider this 2.0-RC1.
20140924 dholland	Make km3 rotate object sizes as originally intended.
20140924 dholland	Make frack print a visible divider when it syncs.
20140922 dholland	Add /bin/tac, which uses unlinked scratch files.
20140919 dholland	Allow giving testbin/bigfile the chunk size to write.
20140919 dholland	Add a design doc about the structure of assignments.
20140919 dholland	Make certain sfsck checks set the exit status properly.
20140918 dholland	Provide general-purpose metadata I/O function in sfs.
20140916 dholland	Remove size workaround in bigexec; fixed solution set.
20140904 dholland	Fix problem with stray symlinks in $(OSTREE)/include.
20140904 dholland	Change sfs_vnode->sv_v to sv_absvn for consistency.
20140904 dholland	Add bloat test; it uses all available memory rapidly.
20140904 dholland	Fix testbin/crash for gcc 4.8.
20140904 dholland	Make testbin/filetest do something useful if no argv.
20140904 dholland	Declare userland err* __DEAD.
20140904 dholland	Add sys/cdefs.h and move userland __DEAD there.
20140903 dholland	Simplify ram.c interface as suggested by my students.
20140829 dholland	Add km4: a kmalloc test for multipage allocations.
20140829 dholland	Add multiexec test for lots of procs in exec at once.
20140828 dholland	Tighten asm constraints for gcc 4.8.
20140828 dholland	Fix inlining for gcc 4.8, and a few other build issues.
20140825 dholland	Remove vnode open count and VOP_INC/DECOPEN.
20140825 dholland	Remove VOP_LASTCLOSE; nothing uses it any more.
20140825 dholland	Don't fsync in sfs_lastclose; it's outdated practice.
20140825 dholland	Remove DEVOP_LASTCLOSE; nothing uses it and we will
........         	never add rewind-on-close tape devices.
20140825 dholland	Use a separate spinlock for vnode refcounts. (Using
........                vfs_biglock causes sleeping while holding spinlocks.)
20140825 dholland	Don't use sfs_io() for directory entries.
20140821 dholland	Replace VOP_TRYSEEK with static VOP_ISSEEKABLE.
20140821 dholland	Make devices check seek position validity on the fly.
20140806 dholland	Make sfs_link reject directories. From Anne Madoff.
20140730 dholland	Make testbin/hog run longer.
20140729 dholland	Document psort's sizing knobs.
20140729 dholland	Make psort big enough to be useful as a fs test.
20140729 dholland	Add ARRAYCOUNT() macro for static array length.
20140729 dholland	Add array_preallocate().
20140729 dholland	Make npages argument of alloc_kpages() unsigned.
20140729 dholland	Tidy up construction/destruction of struct sfs_fs.
20140728 dholland	Make sfs_read/writeblock take the buffer length.
........		(Always do that, even if it's the same everywhere.)
20140728 dholland	struct sfs_super -> struct sfs_superblock.
20140728 dholland	Adjust SFS code to make it more readily extensible.
20140728 dholland	Always call SFS's free block bitmap the "freemap".
20140728 dholland	Rename a bunch of the SFS constants for clarity.
20140725 dholland	Add usemtest for checking the semfs semaphores.
20140724 dholland	Make failed SFS writes not increase the file size.
20140724 dholland	Distinguish MIPS 512M RAM limit and LAMEbus 508M limit.
20140722 dholland	sfs_dir -> sfs_direntry, by popular demand.
20140722 dholland	Add some notes about the atomicity of dup2().
20140722 dholland	Add notes about the limits of syscall atomicity in
........         	multithreaded processes. From Jared Pochtar.
20140722 dholland	Add redirect test to check stdin/stdout redirection.
20140722 dholland	Fix emufs handling of seek positions beyond 2^32.
20140722 dholland	Add bigseek test for checking seeks beyond 2^32.
20140721 dholland	Add file open mode checks to badcall.
20140721 dholland	Fix interaction of kmalloc guards and kmalloc labels.
20140721 dholland	Be more conservative about panic on stray interrupt.
20140721 dholland	Fix userland build to not rebuild unnecessarily.
20140721 dholland	Fix userland build to not belch on removed .h files.
20140721 dholland	In badcall, don't shout if mkdir isn't implemented.
20140717 dholland	Fix misleading comments and variable names in proc.c.
20140716 dholland	Add -w option to parallelvm to wait for forking.
20140716 dholland	Replace a busywait in badcall with the new semaphores.
20140716 dholland	Provide userlevel semaphores via semfs filesystem.
........                Open "sem:<name>", then read to P() and write to V().
20140716 dholland	Move the just-fail vnode op stubs to VFS.
20140710 dholland	Added new test program: sbrktest.
20140710 dholland	Make the libc qsort() actually quicksort.
20140709 dholland	Ship qsort() in libc, not in sfsck's compat code.
20140709 dholland	SWAP{S,L,LL} -> SWAP{16,32,64} in sfs tools.
20140709 dholland	Fix reversed found/expected printout in frack check.
20140518 dholland	Fix embarrasing sort bugs in native sfsck.
20140430 dholland	Add poisondisk tool for testing file system recovery.
20140428 dholland	Add hostcompat logic for 64-bit byte-swapping.
20140417 dholland	Fix zero to use page-sized sbrk calls.
20140415 dholland	Rewrite dumpsfs and make it much more useful.
20140414 dholland	Fix missing initialization of cpu->c_spinlocks.
20140410 dholland	Fix bug in malloc changes. Add assertion to malloctest.
20140326 dholland	Make userlevel malloc allocate in page-sized chunks.
20140326 dholland	Fix the (unused) 64-bit userlevel malloc code.
20140314 dholland	Make forktest show how much output is expected.
20140220 dholland	Add new sy4 test for CVs.
20140211 dholland	Tidy some logic in thread_make_runnable.
20140211 dholland	Mark threads READY when waking them. From Anne Madoff.

kern/main/menu.c
/*
 * Copyright (c) 2000, 2001, 2002, 2003, 2004, 2005, 2008, 2009
 *	The President and Fellows of Harvard College.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE UNIVERSITY AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE UNIVERSITY OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <types.h>
#include <kern/errno.h>
#include <kern/reboot.h>
#include <kern/unistd.h>
#include <limits.h>
#include <lib.h>
#include <uio.h>
#include <clock.h>
#include <thread.h>
#include <proc.h>
#include <vfs.h>
#include <sfs.h>
#include <syscall.h>
#include <test.h>
#include "opt-sfs.h"
#include "opt-net.h"

/*
 * In-kernel menu and command dispatcher.
 */

#define _PATH_SHELL "/bin/sh"

#define MAXMENUARGS  16

////////////////////////////////////////////////////////////
//
// Command menu functions

/*
 * Function for a thread that runs an arbitrary userlevel program by
 * name.
 *
 * Note: this cannot pass arguments to the program. You may wish to
 * change it so it can, because that will make testing much easier
 * in the future.
 *
 * It copies the program name because runprogram destroys the copy
 * it gets by passing it to vfs_open().
 */
static
void
cmd_progthread(void *ptr, unsigned long nargs)
{
	char **args = ptr;
	char progname[128];
	int result;

	KASSERT(nargs >= 1);

	if (nargs > 2) {
		kprintf("Warning: argument passing from menu not supported\n");
	}

	/* Hope we fit. */
	KASSERT(strlen(args[0]) < sizeof(progname));

	strcpy(progname, args[0]);

	result = runprogram(progname);
	if (result) {
		kprintf("Running program %s failed: %s\n", args[0],
			strerror(result));
		return;
	}

	/* NOTREACHED: runprogram only returns on error. */
}

/*
 * Common code for cmd_prog and cmd_shell.
 *
 * Note that this does not wait for the subprogram to finish, but
 * returns immediately to the menu. This is usually not what you want,
 * so you should have it call your system-calls-assignment waitpid
 * code after forking.
 *
 * Also note that because the subprogram's thread uses the "args"
 * array and strings, until you do this a race condition exists
 * between that code and the menu input code.
 */
static
int
common_prog(int nargs, char **args)
{
	struct proc *proc;
	int result;

	/* Create a process for the new program to run in. */
	proc = proc_create_runprogram(args[0] /* name */);
	if (proc == NULL) {
		return ENOMEM;
	}

	result = thread_fork(args[0] /* thread name */,
			proc /* new process */,
			cmd_progthread /* thread function */,
			args /* thread arg */, nargs /* thread arg */);
	if (result) {
		kprintf("thread_fork failed: %s\n", strerror(result));
		proc_destroy(proc);
		return result;
	}

	/*
	 * The new process will be destroyed when the program exits...
	 * once you write the code for handling that.
	 */

	return 0;
}

/*
 * Command for running an arbitrary userlevel program.
 */
static
int
cmd_prog(int nargs, char **args)
{
	if (nargs < 2) {
		kprintf("Usage: p program [arguments]\n");
		return EINVAL;
	}

	/* drop the leading "p" */
	args++;
	nargs--;

	return common_prog(nargs, args);
}

/*
 * Command for starting the system shell.
 */
static
int
cmd_shell(int nargs, char **args)
{
	(void)args;
	if (nargs != 1) {
		kprintf("Usage: s\n");
		return EINVAL;
	}

	args[0] = (char *)_PATH_SHELL;

	return common_prog(nargs, args);
}

/*
 * Command for changing directory.
 */
static
int
cmd_chdir(int nargs, char **args)
{
	if (nargs != 2) {
		kprintf("Usage: cd directory\n");
		return EINVAL;
	}

	return vfs_chdir(args[1]);
}

/*
 * Command for printing the current directory.
 */
static
int
cmd_pwd(int nargs, char **args)
{
	char buf[PATH_MAX+1];
	int result;
	struct iovec iov;
	struct uio ku;

	(void)nargs;
	(void)args;

	uio_kinit(&iov, &ku, buf, sizeof(buf)-1, 0, UIO_READ);
	result = vfs_getcwd(&ku);
	if (result) {
		kprintf("vfs_getcwd failed (%s)\n", strerror(result));
		return result;
	}

	/* null terminate */
	buf[sizeof(buf)-1-ku.uio_resid] = 0;

	/* print it */
	kprintf("%s\n", buf);

	return 0;
}

/*
 * Command for running sync.
 */
static
int
cmd_sync(int nargs, char **args)
{
	(void)nargs;
	(void)args;

	vfs_sync();

	return 0;
}

/*
 * Command for doing an intentional panic.
 */
static
int
cmd_panic(int nargs, char **args)
{
	(void)nargs;
	(void)args;

	panic("User requested panic\n");
	return 0;
}

/*
 * Command for shutting down.
 */
static
int
cmd_quit(int nargs, char **args)
{
	(void)nargs;
	(void)args;

	vfs_sync();
	sys_reboot(RB_POWEROFF);
	thread_exit();
	return 0;
}

/*
 * Command for mounting a filesystem.
 */

/* Table of mountable filesystem types. */
static const struct {
	const char *name;
	int (*func)(const char *device);
} mounttable[] = {
#if OPT_SFS
	{ "sfs", sfs_mount },
#endif
};

static
int
cmd_mount(int nargs, char **args)
{
	char *fstype;
	char *device;
	unsigned i;

	if (nargs != 3) {
		kprintf("Usage: mount fstype device:\n");
		return EINVAL;
	}

	fstype = args[1];
	device = args[2];

	/* Allow (but do not require) colon after device name */
	if (device[strlen(device)-1]==':') {
		device[strlen(device)-1] = 0;
	}

	for (i=0; i<ARRAYCOUNT(mounttable); i++) {
		if (!strcmp(mounttable[i].name, fstype)) {
			return mounttable[i].func(device);
		}
	}
	kprintf("Unknown filesystem type %s\n", fstype);
	return EINVAL;
}

static
int
cmd_unmount(int nargs, char **args)
{
	char *device;

	if (nargs != 2) {
		kprintf("Usage: unmount device:\n");
		return EINVAL;
	}

	device = args[1];

	/* Allow (but do not require) colon after device name */
	if (device[strlen(device)-1]==':') {
		device[strlen(device)-1] = 0;
	}

	return vfs_unmount(device);
}

/*
 * Command to set the "boot fs".
 *
 * The boot filesystem is the one that pathnames like /bin/sh with
 * leading slashes refer to.
 *
 * The default bootfs is "emu0".
 */
static
int
cmd_bootfs(int nargs, char **args)
{
	char *device;

	if (nargs != 2) {
		kprintf("Usage: bootfs device\n");
		return EINVAL;
	}

	device = args[1];

	/* Allow (but do not require) colon after device name */
	if (device[strlen(device)-1]==':') {
		device[strlen(device)-1] = 0;
	}

	return vfs_setbootfs(device);
}

static
int
cmd_kheapstats(int nargs, char **args)
{
	(void)nargs;
	(void)args;

	kheap_printstats();

	return 0;
}

static
int
cmd_kheapgeneration(int nargs, char **args)
{
	(void)nargs;
	(void)args;

	kheap_nextgeneration();

	return 0;
}

static
int
cmd_kheapdump(int nargs, char **args)
{
	if (nargs == 1) {
		kheap_dump();
	}
	else if (nargs == 2 && !strcmp(args[1], "all")) {
		kheap_dumpall();
	}
	else {
		kprintf("Usage: khdump [all]\n");
	}

	return 0;
}

////////////////////////////////////////
//
// Menus.

static
void
showmenu(const char *name, const char *x[])
{
	int ct, half, i;

	kprintf("\n");
	kprintf("%s\n", name);

	for (i=ct=0; x[i]; i++) {
		ct++;
	}
	half = (ct+1)/2;

	for (i=0; i<half; i++) {
		kprintf("    %-36s", x[i]);
		if (i+half < ct) {
			kprintf("%s", x[i+half]);
		}
		kprintf("\n");
	}

	kprintf("\n");
}

static const char *opsmenu[] = {
	"[s]       Shell                     ",
	"[p]       Other program             ",
	"[mount]   Mount a filesystem        ",
	"[unmount] Unmount a filesystem      ",
	"[bootfs]  Set \"boot\" filesystem     ",
	"[pf]      Print a file              ",
	"[cd]      Change directory          ",
	"[pwd]     Print current directory   ",
	"[sync]    Sync filesystems          ",
	"[panic]   Intentional panic         ",
	"[q]       Quit and shut down        ",
	NULL
};

static
int
cmd_opsmenu(int n, char **a)
{
	(void)n;
	(void)a;

	showmenu("OS/161 operations menu", opsmenu);
	return 0;
}

static const char *testmenu[] = {
	"[at]  Array test                    ",
	"[bt]  Bitmap test                   ",
	"[tlt] Threadlist test               ",
	"[km1] Kernel malloc test            ",
	"[km2] kmalloc stress test           ",
	"[km3] Large kmalloc test            ",
	"[km4] Multipage kmalloc test        ",
	"[tt1] Thread test 1                 ",
	"[tt2] Thread test 2                 ",
	"[tt3] Thread test 3                 ",
#if OPT_NET
	"[net] Network test                  ",
#endif
	"[sy1] Semaphore test                ",
	"[sy2] Lock test             (1)     ",
	"[sy3] CV test               (1)     ",
	"[sy4] CV test #2            (1)     ",
	"[fs1] Filesystem test               ",
	"[fs2] FS read stress                ",
	"[fs3] FS write stress               ",
	"[fs4] FS write stress 2             ",
	"[fs5] FS long stress                ",
	"[fs6] FS create stress              ",
	NULL
};

static
int
cmd_testmenu(int n, char **a)
{
	(void)n;
	(void)a;

	showmenu("OS/161 tests menu", testmenu);
	kprintf("    (1) These tests will fail until you finish the "
		"synch assignment.\n");
	kprintf("\n");

	return 0;
}

static const char *mainmenu[] = {
	"[?o] Operations menu                ",
	"[?t] Tests menu                     ",
	"[kh] Kernel heap stats              ",
	"[khgen] Next kernel heap generation ",
	"[khdump] Dump kernel heap           ",
	"[q] Quit and shut down              ",
	NULL
};

static
int
cmd_mainmenu(int n, char **a)
{
	(void)n;
	(void)a;

	showmenu("OS/161 kernel menu", mainmenu);
	return 0;
}

////////////////////////////////////////
//
// Command table.

static struct {
	const char *name;
	int (*func)(int nargs, char **args);
} cmdtable[] = {
	/* menus */
	{ "?",		cmd_mainmenu },
	{ "h",		cmd_mainmenu },
	{ "help",	cmd_mainmenu },
	{ "?o",		cmd_opsmenu },
	{ "?t",		cmd_testmenu },

	/* operations */
	{ "s",		cmd_shell },
	{ "p",		cmd_prog },
	{ "mount",	cmd_mount },
	{ "unmount",	cmd_unmount },
	{ "bootfs",	cmd_bootfs },
	{ "pf",		printfile },
	{ "cd",		cmd_chdir },
	{ "pwd",	cmd_pwd },
	{ "sync",	cmd_sync },
	{ "panic",	cmd_panic },
	{ "q",		cmd_quit },
	{ "exit",	cmd_quit },
	{ "halt",	cmd_quit },

	/* stats */
	{ "kh",         cmd_kheapstats },
	{ "khgen",      cmd_kheapgeneration },
	{ "khdump",     cmd_kheapdump },

	/* base system tests */
	{ "at",		arraytest },
	{ "bt",		bitmaptest },
	{ "tlt",	threadlisttest },
	{ "km1",	kmalloctest },
	{ "km2",	kmallocstress },
	{ "km3",	kmalloctest3 },
	{ "km4",	kmalloctest4 },
#if OPT_NET
	{ "net",	nettest },
#endif
	{ "tt1",	threadtest },
	{ "tt2",	threadtest2 },
	{ "tt3",	threadtest3 },
	{ "sy1",	semtest },

	/* synchronization assignment tests */
	{ "sy2",	locktest },
	{ "sy3",	cvtest },
	{ "sy4",	cvtest2 },

	/* file system assignment tests */
	{ "fs1",	fstest },
	{ "fs2",	readstress },
	{ "fs3",	writestress },
	{ "fs4",	writestress2 },
	{ "fs5",	longstress },
	{ "fs6",	createstress },

	{ NULL, NULL }
};

/*
 * Process a single command.
 */
static
int
cmd_dispatch(char *cmd)
{
	struct timespec before, after, duration;
	char *args[MAXMENUARGS];
	int nargs=0;
	char *word;
	char *context;
	int i, result;

	for (word = strtok_r(cmd, " \t", &context);
	     word != NULL;
	     word = strtok_r(NULL, " \t", &context)) {

		if (nargs >= MAXMENUARGS) {
			kprintf("Command line has too many words\n");
			return E2BIG;
		}
		args[nargs++] = word;
	}

	if (nargs==0) {
		return 0;
	}

	for (i=0; cmdtable[i].name; i++) {
		if (*cmdtable[i].name && !strcmp(args[0], cmdtable[i].name)) {
			KASSERT(cmdtable[i].func!=NULL);

			gettime(&before);

			result = cmdtable[i].func(nargs, args);

			gettime(&after);
			timespec_sub(&after, &before, &duration);

			kprintf("Operation took %llu.%09lu seconds\n",
				(unsigned long long) duration.tv_sec,
				(unsigned long) duration.tv_nsec);

			return result;
		}
	}

	kprintf("%s: Command not found\n", args[0]);
	return EINVAL;
}

/*
 * Evaluate a command line that may contain multiple semicolon-delimited
 * commands.
 *
 * If "isargs" is set, we're doing command-line processing; print the
 * comamnds as we execute them and panic if the command is invalid or fails.
 */
static
void
menu_execute(char *line, int isargs)
{
	char *command;
	char *context;
	int result;

	for (command = strtok_r(line, ";", &context);
	     command != NULL;
	     command = strtok_r(NULL, ";", &context)) {

		if (isargs) {
			kprintf("OS/161 kernel: %s\n", command);
		}

		result = cmd_dispatch(command);
		if (result) {
			kprintf("Menu command failed: %s\n", strerror(result));
			if (isargs) {
				panic("Failure processing kernel arguments\n");
			}
		}
	}
}

/*
 * Command menu main loop.
 *
 * First, handle arguments passed on the kernel's command line from
 * the bootloader. Then loop prompting for commands.
 *
 * The line passed in from the bootloader is treated as if it had been
 * typed at the prompt. Semicolons separate commands; spaces and tabs
 * separate words (command names and arguments).
 *
 * So, for instance, to mount an SFS on lhd0 and make it the boot
 * filesystem, and then boot directly into the shell, one would use
 * the kernel command line
 *
 *      "mount sfs lhd0; bootfs lhd0; s"
 */

void
menu(char *args)
{
	char buf[64];

	menu_execute(args, 1);

	while (1) {
		kprintf("OS/161 kernel [? for menu]: ");
		kgets(buf, sizeof(buf));
		menu_execute(buf, 0);
	}
}
